File: site/scripts/generate-audio.py
#!/usr/bin/env python3
"""用 edge-tts（en-GB-SoniaNeural，与 jingyu 一致）生成 850 词 MP3。"""

from __future__ import annotations

import asyncio
import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
OUT = ROOT / "02-vocabulary" / "audio"
WORDS_FILE = ROOT / "02-vocabulary" / "basic-english-850.txt"
VOICE = "en-GB-SoniaNeural"


def slug(word: str) -> str:
    s = re.sub(r"[^a-zA-Z0-9]+", "-", word.lower()).strip("-")
    return s[:60] or "word"


async def gen_one(word: str, out: Path) -> None:
    import edge_tts

    out.parent.mkdir(parents=True, exist_ok=True)
    comm = edge_tts.Communicate(word, VOICE)
    await comm.save(str(out))


async def run(words: list[str], force: bool) -> int:
    n = 0
    for i, w in enumerate(words, 1):
        path = OUT / f"{slug(w)}.mp3"
        if path.exists() and not force:
            continue
        await gen_one(w, path)
        n += 1
        if i % 25 == 0:
            print(f"  … {i}/{len(words)}", flush=True)
    return n


def main() -> None:
    force = "--force" in sys.argv
    if not WORDS_FILE.exists():
        print(f"missing {WORDS_FILE}", file=sys.stderr)
        sys.exit(1)
    words = [ln.strip() for ln in WORDS_FILE.read_text(encoding="utf-8").splitlines() if ln.strip()]
    print(f"voice={VOICE} words={len(words)} out={OUT}")
    n = asyncio.run(run(words, force))
    print(f"generated {n} new mp3")


if __name__ == "__main__":
    main()

File: site/scripts/generate-audio.mjs
#!/usr/bin/env node
/** 用 edge-tts en-GB-SoniaNeural 生成 850 词发音 MP3 */
import { execFile } from "node:child_process";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { promisify } from "node:util";

const execFileAsync = promisify(execFile);
const __dirname = path.dirname(fileURLToPath(import.meta.url));
const root = path.resolve(__dirname, "../..");
const audioDir = path.join(root, "02-vocabulary/audio");
const edgeTts = path.join(__dirname, ".venv-tts/bin/edge-tts");
const VOICE = "en-GB-SoniaNeural";

const words = fs
  .readFileSync(path.join(root, "02-vocabulary/basic-english-850.txt"), "utf8")
  .trim()
  .split(/\s+/);

function slug(word) {
  return word.toLowerCase().replace(/[^a-z0-9]+/g, "-").replace(/^-|-$/g, "").slice(0, 60) || "word";
}

async function generate(word) {
  const file = path.join(audioDir, `${slug(word)}.mp3`);
  if (fs.existsSync(file) && !process.argv.includes("--all")) return "skip";
  await execFileAsync(edgeTts, ["--voice", VOICE, "--text", word, "--write-media", file]);
  return "ok";
}

fs.mkdirSync(audioDir, { recursive: true });

let ok = 0;
let skip = 0;
let fail = 0;

for (const word of words) {
  try {
    const r = await generate(word);
    if (r === "skip") skip++;
    else {
      ok++;
      process.stdout.write(`\r✓ ${ok} generated (${word})`);
    }
  } catch (e) {
    fail++;
    console.error(`\n✗ ${word}:`, e.message);
  }
}

console.log(`\nDone: ${ok} generated, ${skip} skipped, ${fail} failed → ${audioDir}`);
